streams/shared/src/main/scala/zio/stream/ZPipeline.scala

```diff
@@ -1487,6 +1488,117 @@ object ZPipeline extends ZPipelinePlatformSpecificConstructors {
       }
     )
 
+  /**
+   * Decode each pair of hex digit input bytes (both lower or upper case letters
+   * are allowed) as one output byte.
+   */
+  def hexDecode(implicit trace: Trace): ZPipeline[Any, EncodingException, Byte, Byte] = {
```

Comment:
What do you think about failing with a `CharacterCodingException` here? This is what we already do with other decoding pipelines such as `utf8Decode` and it seems like it would be nice to have consistency between the different decoding pipelines as well as avoid introducing our own error type if possible.

Reply:
I like the idea of using an existing exception if an appropriate one is available. Since what we are doing has nothing to do with character sets though, and having read the description of that exception in the JavaDoc for `CharacterCodingException`, it feels like a stretch to me (because this isn't a "character coding" issue):

> Checked exception thrown when a character encoding or decoding error occurs.

There is a subclass `UnmappableCharacterException` for which the JavaDoc reads closer to the semantics here (because it explicitly mentions input and output byte sequences):

> Checked exception thrown when an input character (or byte) sequence is valid but cannot be mapped to an output byte (or character) sequence.

What do you think?

Reply:
Also, the impetus for creating this was a desire to not need to bring in Apache Codec's Hex encoder / decoder when working with ZIO Streams, specifically in testing where I wanted to pass in specific byte sequences. I implemented only UTF-8 lower case on the encode side and UTF-8 mixed case on the decode side. Apache Codec's implementation supports using different character sets but I didn't bother.

See: https://commons.apache.org/proper/commons-codec/apidocs/org/apache/commons/codec/binary/Hex.html

Do you think it would be worthwhile to make this support lower / upper selection on the encode side (default lower)?

Do you think it would be worthwhile to make this support specifying a different character set / encoding (default UTF-8)?

Of the two, it seems to me more like someone would wish upper case encoding instead of lower case than EBCDIC or something.

Reply:
Understood. Logically better of course. I was hesitant because its double the memory size though due to Char being 16 bits. I'll change to Char as the type for input to decode and output of encode.

Reply:
I think at this point the `CharacterCodingException` seems to fit since a character decoding exception has occurred.

Reply:
I can make the change to use CharacterCodingException because it takes the standard message and cause arguments, so at least I can retain my descriptive messages.

Do you intend that I use that exception everywhere I'm using EncodingException currently? If not, that makes the error type problematic. I still don't think they are all legitimately CharacterCodingExceptions but I'm willing to comply with your intent.

Reply:
Yes, my suggestion would be that the error type of the pipeline be `CharacterCodingException` which would mean every error we fail with would be that. I think that is fair since it is for a character decoding error that occurs and these are errors in decoding hexadecimal input bytes into characters. This also has the advantage of using an existing error type and being consistent with other decoding pipelines.

Reply:
@adamgfraser and I had a side discussion about the challenges of using the existing JDK exceptions. Their names seem attractive but their semantics really are not a match for what we are doing, which is closer to what was done with the existing ZIO Streams class CompressionException. We want to preserve the ability to have meaningful exception messages for example, which is not the case with the JDK exceptions discussed above.

I believe the current state of the code on this branch is ready as-is. @adamgfraser is that right?